libs/utils/trace_analysis.py

```diff
@@ -785,3 +787,117 @@ def plotSchedTuneConf(self):
         # Save generated plots into datadir
         figname = '{}/{}schedtune_conf.png'.format(self.plotsdir, self.prefix)
         pl.savefig(figname, bbox_inches='tight')
+
+    def getClusterResidency(self, cluster, cluster_name):
+        """
+        Get a DataFrame with per cluster frequency residency, i.e. amount of
+        time spent at a given frequency in each cluster.
+
+        :param cluster: list of CPU IDs belonging to the cluster
+        :type cluster: list(int)
+
+        :param cluster_name: name of the cluster
+        :type cluster_name: str
+        """
+
+        if not self.trace.hasEvents('cpu_frequency'):
+            logging.warn('Events [cpu_frequency] not found, '\
+                         'plot DISABLED!')
+            return None
+        if not self.trace.hasEvents('cpu_idle'):
+            logging.warn('Events [cpu_idle] not found, '\
+                         'plot DISABLED!')
+            return None
+
+        freq_dfr = self.trace.df('cpu_frequency')
+        cpu_freqs = freq_dfr[freq_dfr.apply(lambda x: x.cpu in cluster,
+                                            axis=1)]
+        idle_dfr = self.trace.df('cpu_idle')
+        cpu_idle = idle_dfr[idle_dfr.apply(lambda x: x.cpu_id in cluster,
+                                           axis=1)]
+
+        # time instants where CPUs go to an idle state
+        idle_times = cpu_idle[cpu_idle['state'] != IDLE_STATE]
+        # time instants where CPUs leave an idle state
+        non_idle_times = cpu_idle[cpu_idle['state'] == IDLE_STATE]
+
+        if idle_times.index[0] < non_idle_times.index[0]:
+            idle_times = idle_times.iloc[1:]
+
+        time_intervals = [t2 - t1 for t1, t2 in zip(non_idle_times.index,
+                                                    idle_times.index)]
```

Comment:
Good point! I think we should implement a `getCPUResidency` because in another situations we want per-CPU info and just get the data from there. What do you think?